### ISSUE TITLE
update GSOC.md and CONTRIBUTING.md

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#15845]
taken from https://github.com/checkstyle/checkstyle/pull/14514

address review items.
each file should be updated in separate PR.

https://github.com/checkstyle/checkstyle/blob/master/.github/GSOC.md

https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md

https://github.com/checkstyle/checkstyle/blob/master/docs/BEGINNING_DEVELOPMENT.md

**Comment by Kushagra-Sedwal** [comment]
Heyy!! I would like to work on this issue

**Comment by mahfouz72** [comment]
> I hope you still remember what was most confusing point to start contribution and succeed in first pull request.

Yes, I believe these new docs will make it much easier for new contributors to get started. I glanced on them and they provide a way clearer explanation of each step involved in contributing and planning for GSoC.

Thanks to @nrmancuso for consistently taking the time to help new contributors.

I’ll review the documents again soon and see if I can spot any additional improvements I can make.

**Comment by romani** [comment]
I fixed as much as I could in referenced PR.

We can always to change more is someone see points to improve and send updates.

If no updates send until February 2025, this issue can be closed.
All updates to doc can always be done without issue, just PR with "doc: " prefix for commit.

**Comment by Zopsss** [comment]
The new CONTRIBUTING.md & GSOC.md looks amazing and it will definitely help many new-comers to get started. I'm currently a bit busy with college and personal stuff but I will take a glance at it once I get some free time. 

Btw thanks for the awesome work @nrmancuso 🚀
